### PR TITLE
upgrade pylint version to 2.14.4

### LINKED ISSUE #6200: Upgrade `pylint` to  `2.14.4`
We currently use pylint 2.11.1. This is a bit old so we should upgrade it to the latest version (2.14.4):

https://github.com/mlflow/mlflow/blob/d40780be361f4bd2741c2e8fcbd428c1d693edcf/requirements/lint-requirements.txt#L1

- https://pypi.org/project/pylint/2.11.1/
- https://pypi.org/project/pylint/2.14.4/

### PR DESCRIPTION
**Pull request by KarthikKothareddy**
Signed-off-by: Karthik Kothareddy <karthikreddy.kp@gmail.com>

## Related Issues/PRs

Close #6200 

## What changes are proposed in this pull request?
The change includes an update to the pylint version to 2.14.4 from 2.11.1

(Please fill in changes proposed in this fix)

## How is this patch tested?

- [ ] I have written tests (not required for typo or doc fix) and confirmed the proposed feature/bug-fix/change works.

## Does this PR change the documentation?
No

- [x] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly by following the steps below.

1. Check the status of the `ci/circleci: build_doc` check. If it's successful, proceed to the
   next step, otherwise fix it.
2. Click `Details` on the right to open the job page of CircleCI.
3. Click the `Artifacts` tab.
4. Click `docs/build/html/index.html`.
5. Find the changed pages / sections and make sure they render correctly.

### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?
Components
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/pipelines`: Pipelines, Pipeline APIs, Pipeline configs, Pipeline Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

**Comment by harupy** [comment]
@KarthikKothareddy Can you try running `./dev/lint.sh` on your machine? I this pylint should report some lint errors (it did on my machine). Can you fix them?

**Comment by harupy** [comment]
@KarthikKothareddy Can I push a commit to fix this test error?

https://github.com/mlflow/mlflow/runs/7243099478?check_suite_focus=true#step:7:71

```
E       AttributeError: module 'pylint.testutils' has no attribute 'Message'
```

**Comment by KarthikKothareddy** [comment]
@harupy Of course please feel free to commit.

**Comment by harupy** [comment]
```
************* Module /home/runner/work/mlflow/mlflow/pylintrc
Error: pylintrc:1:0: E0015: Unrecognized option found: no-space-check (unrecognized-option)
Error: pylintrc:1:0: W0012: Unknown option value for '--disable', expected a valid pylint message and got 'long-suffix' (unknown-option-value)
Error: pylintrc:1:0: W0012: Unknown option value for '--disable', expected a valid pylint message and got 'old-ne-operator' (unknown-option-value)
Error: pylintrc:1:0: W0012: Unknown option value for '--disable', expected a valid pylint message and got 'old-octal-literal' (unknown-option-value)
Error: pylintrc:1:0: W0012: Unknown option value for '--disable', expected a valid pylint message and got 'non-ascii-bytes-literal' (unknown-option-value)
Error: pylintrc:1:0: W0012: Unknown option value for '--disable', expected a valid pylint message and got 'locally-enabled' (unknown-option-value)
Error: pylintrc:1:0: W0012: Unknown option value for '--disable', expected a valid pylint message and got 'eq-without-hash' (unknown-option-value)
************* Module mlflow.utils.autologging_utils.logging_and_warnings
Error: mlflow/utils/autologging_utils/logging_and_warnings.py:65:0: W1404: Implicit string concatenation found in call (implicit-str-concat)
************* Module tests.autologging.fixtures
Error: tests/autologging/fixtures.py:46:11: E0601: Using variable 'prev_env_var_value' before assignment (used-before-assignment)
Error: tests/autologging/fixtures.py:59:11: E0601: Using variable 'prev_env_var_value' before assignment (used-before-assignment)
************* Module tests.autologging.test_autologging_safety_unit
Error: tests/autologging/test_autologging_safety_unit.py:155:42: E0601: Using variable 'prev_logger' before assignment (used-before-assignment)
Error: tests/autologging/test_autologging_safety_unit.py:169:11: E0601: Using variable 'prev_env_var_value' before assignment (used-before-assignment)
************* Module test_shap
Error: tests/shap/test_shap.py:189:15: E0601: Using variable 'explanation_uri' before assignment (used-before-assignment)
************* Module tests.store.tracking.test_sqlalchemy_store
Error: tests/store/tracking/test_sqlalchemy_store.py:[22](https://github.com/mlflow/mlflow/runs/7245729527?check_suite_focus=true#step:8:23)28:0: W1404: Implicit string concatenation found in call (implicit-str-concat)
************* Module test_fluent
Error: tests/tracking/fluent/test_fluent.py:[35](https://github.com/mlflow/mlflow/runs/7245729527?check_suite_focus=true#step:8:36)7:8: E0601: Using variable 'process' before assignment (used-before-assignment)
```

**Comment by harupy** [comment]
I think these options no longer exist in pylint 2.14.4. Let's remove them.

```
Error: pylintrc:1:0: E0015: Unrecognized option found: no-space-check (unrecognized-option)
Error: pylintrc:1:0: W0012: Unknown option value for '--disable', expected a valid pylint message and got 'long-suffix' (unknown-option-value)
Error: pylintrc:1:0: W0012: Unknown option value for '--disable', expected a valid pylint message and got 'old-ne-operator' (unknown-option-value)
Error: pylintrc:1:0: W0012: Unknown option value for '--disable', expected a valid pylint message and got 'old-octal-literal' (unknown-option-value)
Error: pylintrc:1:0: W0012: Unknown option value for '--disable', expected a valid pylint message and got 'non-ascii-bytes-literal' (unknown-option-value)
Error: pylintrc:1:0: W0012: Unknown option value for '--disable', expected a valid pylint message and got 'locally-enabled' (unknown-option-value)
Error: pylintrc:1:0: W0012: Unknown option value for '--disable', expected a valid pylint message and got 'eq-without-hash' (unknown-option-value)
```

**Comment by harupy** [comment]
```
Error: tests/autologging/fixtures.py:46:11: E0601: Using variable 'prev_env_var_value' before assignment (used-before-assignment)
Error: tests/autologging/fixtures.py:59:11: E0601: Using variable 'prev_env_var_value' before assignment (used-before-assignment)
```

https://github.com/mlflow/mlflow/blob/f4f2d6a228f84e9bd64d65cf52af78192e90d1f2/tests/autologging/fixtures.py#L38-L50

When `os.environ.pop` fails (it usually doesn't), line 46 fails because `prev_env_var_value` is not defined yet. I think we can move `prev_env_var_value = os.environ.pop(_AUTOLOGGING_TEST_MODE_ENV_VAR, None)` outside of try-catch.

**Comment by KarthikKothareddy** [comment]
> @KarthikKothareddy Can you fix the following errors?
> 
> ```
> ************* Module /home/runner/work/mlflow/mlflow/pylintrc
> Error: pylintrc:1:0: E0015: Unrecognized option found: no-space-check (unrecognized-option)
> Error: pylintrc:1:0: W0012: Unknown option value for '--disable', expected a valid pylint message and got 'long-suffix' (unknown-option-value)
> Error: pylintrc:1:0: W0012: Unknown option value for '--disable', expected a valid pylint message and got 'old-ne-operator' (unknown-option-value)
> Error: pylintrc:1:0: W0012: Unknown option value for '--disable', expected a valid pylint message and got 'old-octal-literal' (unknown-option-value)
> Error: pylintrc:1:0: W0012: Unknown option value for '--disable', expected a valid pylint message and got 'non-ascii-bytes-literal' (unknown-option-value)
> Error: pylintrc:1:0: W0012: Unknown option value for '--disable', expected a valid pylint message and got 'locally-enabled' (unknown-option-value)
> Error: pylintrc:1:0: W0012: Unknown option value for '--disable', expected a valid pylint message and got 'eq-without-hash' (unknown-option-value)
> ************* Module mlflow.utils.autologging_utils.logging_and_warnings
> Error: mlflow/utils/autologging_utils/logging_and_warnings.py:65:0: W1404: Implicit string concatenation found in call (implicit-str-concat)
> ************* Module tests.autologging.fixtures
> Error: tests/autologging/fixtures.py:46:11: E0601: Using variable 'prev_env_var_value' before assignment (used-before-assignment)
> Error: tests/autologging/fixtures.py:59:11: E0601: Using variable 'prev_env_var_value' before assignment (used-before-assignment)
> ************* Module tests.autologging.test_autologging_safety_unit
> Error: tests/autologging/test_autologging_safety_unit.py:155:42: E0601: Using variable 'prev_logger' before assignment (used-before-assignment)
> Error: tests/autologging/test_autologging_safety_unit.py:169:11: E0601: Using variable 'prev_env_var_value' before assignment (used-before-assignment)
> ************* Module test_shap
> Error: tests/shap/test_shap.py:189:15: E0601: Using variable 'explanation_uri' before assignment (used-before-assignment)
> ************* Module tests.store.tracking.test_sqlalchemy_store
> Error: tests/store/tracking/test_sqlalchemy_store.py:[22](https://github.com/mlflow/mlflow/runs/7245729527?check_suite_focus=true#step:8:23)28:0: W1404: Implicit string concatenation found in call (implicit-str-concat)
> ************* Module test_fluent
> Error: tests/tracking/fluent/test_fluent.py:[35](https://github.com/mlflow/mlflow/runs/7245729527?check_suite_focus=true#step:8:36)7:8: E0601: Using variable 'process' before assignment (used-before-assignment)
> ```
> 
> Let me know if you have any questions. I'm happy to help.

@harupy yeah sure. I have some time this evening and I plan to fix all the linting issues along with the ones you mentioned.

File: dev/generate_requirements.py [hunk 1]
```diff
@@ -18,6 +18,10 @@
 
 
 def parse_args(args):
+    """
+    Returns parsed arguments
+    :param args: Arguments to parse
+    """
```
Comment (harupy) [inline dev/generate_requirements.py:24]: @KarthikKothareddy Why do we need this change?

Comment (KarthikKothareddy) [reply dev/generate_requirements.py:24]: @harupy pylint 2.14.4 triggered warnings about methods with no docstrings in the ./dev directory.

Comment (harupy) [reply dev/generate_requirements.py:24]: The category of `missing-doctsring` is C and it's disabled in this repo:
https://github.com/mlflow/mlflow/blob/f4f2d6a228f84e9bd64d65cf52af78192e90d1f2/pylintrc#L143

Comment (harupy) [reply dev/generate_requirements.py:24]: @KarthikKothareddy How did you run `pylint`? Did you run it in the repository root?

Comment (harupy) [reply dev/generate_requirements.py:24]: You can just fix the errors in https://github.com/mlflow/mlflow/pull/6213#issuecomment-1178550661.

Comment (KarthikKothareddy) [reply dev/generate_requirements.py:24]: I ran them from within 'dev' directory. I just realized that the 'lint.sh' should have been run on root directory, am I correct?

Comment (harupy) [reply dev/generate_requirements.py:24]: yes, it must be run on the root directory.

Comment (KarthikKothareddy) [reply dev/generate_requirements.py:24]: Should I go back to previous commit without the 'changes' to the dev directory scripts and then take it from there?

Comment (harupy) [reply dev/generate_requirements.py:24]: yes

**Comment by harupy** [comment]
@KarthikKothareddy Can you also update `pylintrc` to remove the unknown/unrecognized option?

```
Error: pylintrc:1:0: E0015: Unrecognized option found: no-space-check (unrecognized-option)
Error: pylintrc:1:0: W0012: Unknown option value for '--disable', expected a valid pylint message and got 'long-suffix' (unknown-option-value)
Error: pylintrc:1:0: W0012: Unknown option value for '--disable', expected a valid pylint message and got 'old-ne-operator' (unknown-option-value)
Error: pylintrc:1:0: W0012: Unknown option value for '--disable', expected a valid pylint message and got 'old-octal-literal' (unknown-option-value)
Error: pylintrc:1:0: W0012: Unknown option value for '--disable', expected a valid pylint message and got 'non-ascii-bytes-literal' (unknown-option-value)
Error: pylintrc:1:0: W0012: Unknown option value for '--disable', expected a valid pylint message and got 'locally-enabled' (unknown-option-value)
Error: pylintrc:1:0: W0012: Unknown option value for '--disable', expected a valid pylint message and got 'eq-without-hash' (unknown-option-value)
```

**Comment by KarthikKothareddy** [comment]
> Error: pylintrc:1:0: E0015: Unrecognized option found: no-space-check

Sure will, just to confirm the approach will be to remove the `no-space-check` altogether and then all the Unknown options.

**Comment by harupy** [comment]
> just to confirm the approach will be to remove the no-space-check altogether and then all the Unknown options.

I think it's ok.

**Comment by KarthikKothareddy** [comment]
@harupy I also these issues on my machine, is it okay if I address them (mostly related to unused function arguments)

`************* Module mlflow.gluon._autolog
mlflow/gluon/_autolog.py:14:0: W0613: Unused argument 'args' (unused-argument)
mlflow/gluon/_autolog.py:14:0: W0613: Unused argument 'kwargs' (unused-argument)
mlflow/gluon/_autolog.py:25:0: W0613: Unused argument 'args' (unused-argument)
mlflow/gluon/_autolog.py:25:0: W0613: Unused argument 'kwargs' (unused-argument)
mlflow/gluon/_autolog.py:37:0: W0613: Unused argument 'args' (unused-argument)
mlflow/gluon/_autolog.py:37:0: W0613: Unused argument 'kwargs' (unused-argument)
************* Module mlflow.paddle._paddle_autolog
mlflow/paddle/_paddle_autolog.py:38:29: W0613: Unused argument 'logs' (unused-argument)
mlflow/paddle/_paddle_autolog.py:46:27: W0613: Unused argument 'logs' (unused-argument)
************* Module mlflow.pytorch._pytorch_autolog
mlflow/pytorch/_pytorch_autolog.py:212:38: W0613: Unused argument 'pl_module' (unused-argument)
mlflow/pytorch/_pytorch_autolog.py:242:27: W0613: Unused argument 'trainer' (unused-argument)
mlflow/pytorch/_pytorch_autolog.py:242:36: W0613: Unused argument 'pl_module' (unused-argument)
mlflow/pytorch/_pytorch_autolog.py:254:35: W0613: Unused argument 'pl_module' (unused-argument)`

File: tests/spark/autologging/ml/test_pyspark_ml_autologging.py [hunk 2]
```diff
@@ -389,7 +389,7 @@ def setModel(self, model: SparkModel):
         def setEvaluator(self, evaluator: Evaluator):
             return self._set(evaluator=evaluator)
 
-        def _fit(self, dataset):
+        def _fit(self):
```
Comment (harupy) [inline tests/spark/autologging/ml/test_pyspark_ml_autologging.py:392]: Can we revert this change?

File: tests/store/tracking/test_sqlalchemy_store.py [hunk 3]
```diff
@@ -2225,7 +2225,8 @@ def test_search_runs_correctly_filters_large_data(self):
 
         run_results = self.store.search_runs(
             [experiment_id],
-            "metrics.mkey_0 < 26 and metrics.mkey_0 > 5 " "and tags.tkey_0 = 'tval_0' ",
+            "metrics.mkey_0 < 26 and metrics.mkey_0 > 5 " 
+            "and tags.tkey_0 = 'tval_0' ",
```
Comment (harupy) [inline tests/store/tracking/test_sqlalchemy_store.py:2229]:
```suggestion
            "metrics.mkey_0 < 26 and metrics.mkey_0 > 5 and tags.tkey_0 = 'tval_0' ",
```

**Comment by harupy** [comment]
> @harupy I also saw these issues pop up on my machine, is it okay if I address them (mostly related to unused function arguments)
> 
> ************* Module mlflow.gluon._autolog mlflow/gluon/_autolog.py:14:0: W0613: Unused argument 'args' (unused-argument) mlflow/gluon/_autolog.py:14:0: W0613: Unused argument 'kwargs' (unused-argument) mlflow/gluon/_autolog.py:25:0: W0613: Unused argument 'args' (unused-argument) mlflow/gluon/_autolog.py:25:0: W0613: Unused argument 'kwargs' (unused-argument) mlflow/gluon/_autolog.py:37:0: W0613: Unused argument 'args' (unused-argument) mlflow/gluon/_autolog.py:37:0: W0613: Unused argument 'kwargs' (unused-argument) ************* Module mlflow.paddle._paddle_autolog mlflow/paddle/_paddle_autolog.py:38:29: W0613: Unused argument 'logs' (unused-argument) mlflow/paddle/_paddle_autolog.py:46:27: W0613: Unused argument 'logs' (unused-argument) ************* Module mlflow.pytorch._pytorch_autolog mlflow/pytorch/_pytorch_autolog.py:212:38: W0613: Unused argument 'pl_module' (unused-argument) mlflow/pytorch/_pytorch_autolog.py:242:27: W0613: Unused argument 'trainer' (unused-argument) mlflow/pytorch/_pytorch_autolog.py:242:36: W0613: Unused argument 'pl_module' (unused-argument) mlflow/pytorch/_pytorch_autolog.py:254:35: W0613: Unused argument 'pl_module' (unused-argument)

I guess pytorch, paddle, and mxnet are not installed in your machine.

**Comment by harupy** [comment]
@KarthikKothareddy Can you fix the DCO check failure? https://github.com/mlflow/mlflow/pull/6213/checks?check_run_id=7308561797

**Review by harupy — APPROVED** [review]
@KarthikKothareddy Thanks for the contribution! LGTM :)
